### PR TITLE
[react-native] Support arrays in view style props.

### LINKED ISSUE #17308: react-native View / Text / ScrollView lost the ability to use arrays in style prop
- [x] I tried using the `@types/react-native` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).

Can't assign an array to style (and it's permitted by the docs in https://facebook.github.io/react-native/docs/style.html)

image

Don't know when the change took place. The index.d.ts now shows

image

Should be either `ViewStyle | ViewStyle[];` or have a helper type generic (since it's common to a lot of components that use styles, like TextStyle, FlexStyle, etc)

### PR DESCRIPTION
**Pull request by newyankeecodeshop**
Fixes #17308
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://facebook.github.io/react-native/docs/style.html
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Note that the array uses `| undefined` because React will ignore undefined values in the array. This is also indicated by the flow type RN uses for stylesheet props:
https://github.com/facebook/react-native/blob/master/Libraries/StyleSheet/StyleSheetTypes.js

### CODE REVIEW

**Comment by iRoachie** [comment]
This is already possible. What's the reason for this PR?

Example below shows no errors
```jsx
<View style={[styles.item, { backgroundColor: props.color }]}>
```

**Comment by huhuanming** [comment]
`Array<StyleProps>` and` StyleProps[]`, which is better one?
TSLint rules recommend using `StyleProps[]`.

**Comment by newyankeecodeshop** [comment]
@iRoachie What version of TypeScript are you using? When I upgraded a project to 2.4.1, I started getting these errors.

**Comment by newyankeecodeshop** [comment]
I used `Array<>` because I thought it would look better than `(ViewStyle | undefined)[]`.

**Comment by iRoachie** [comment]
@newyankeecodeshop Ahh 2.4 just got released yesterday, and the typings currently if you check at the top are for Typescript 2.3. So haven't tested as yet. Will upgrade mine now and see

**Comment by iRoachie** [comment]
@newyankeecodeshop @huhuanming Confirming that they are broken in TypeScript 2.4

File: types/react-native/index.d.ts [hunk 1]
```diff
@@ -891,7 +891,7 @@ export interface TextProperties extends TextPropertiesIOS, TextPropertiesAndroid
     /**
      * @see https://facebook.github.io/react-native/docs/text.html#style
      */
-    style?: TextStyle
+    style?: TextStyle | Array<TextStyle | undefined>
```
Comment (iRoachie) [inline types/react-native/index.d.ts]: Can just use `TextStyle | TextStyle[]`

**Review by iRoachie — CHANGES_REQUESTED** [review]
Change all the Array<> to the Type[]. What's the purpose for the undefined?

**Comment by iRoachie** [comment]
All the style props are broken, may want to make a generic to handle all of them. 
`type StyleProp<T> = T | T[];`

**Comment by alloy** [comment]
Could you add some test coverage as well?

**Comment by iRoachie** [comment]
@alloy Do you have anyway we could do this? Setting the TS version to 2.4 doesn't work for dslint. #17626

**Comment by alloy** [comment]
Yeah, my bad, I didn’t mean for it to fail CI, so no need to check that version change in, but make sure it’s a test that would fail locally when using TS 2.4.

File: types/react-native/index.d.ts (hunk 1 above)
Comment (newyankeecodeshop) [inline types/react-native/index.d.ts]: The flow type supports undefined values in the array, and that would work better for optionals where a given prop might be `ViewStyle?`. If you want to put that in an array, the array would need to support undefined values, right?

Comment (iRoachie) [inline types/react-native/index.d.ts]: Can you give an example where undefined would appear in that array?

Comment (newyankeecodeshop) [inline types/react-native/index.d.ts]: An example of why the array needs to support undefined values:
```
interface MyComponentProps {
   extraStyle?: ViewStyle
}

function MyComponent(props: MyComponentProps) {
  const styles = [props.style, props.extraStyle];
  // styles is of type "(ViewStyle | undefined)[]"
  return <View style={styles} />;
}
```

**Comment by newyankeecodeshop** [comment]
Does anyone know why this hasn't caused a problem before? It seems like passing an array for `style` should not have worked, since arrays don't implement the `ViewStyle` interface.

File: types/react-native/index.d.ts (hunk 1 above)
Comment (petejkim) [inline types/react-native/index.d.ts]: :+1:

**Comment by huhuanming** [comment]
In TS 2.4
`<View style={[styles.first, styles.second] as ViewStyle}>`
Or just need a better way.

**Review by iRoachie — APPROVED** [review]

**Comment by petejkim** [comment]
Are we merging this?

**Comment by alloy** [comment]
@iRoachie @huhuanming in the future send me a ping when you’ve approved a PR, otherwise I may miss it.

**Comment by huhuanming** [comment]
fine 👍

**Comment by pocesar** [comment]
missing TextStyle as well

**Comment by dannycochran** [comment]
Would be good to get this landed -- Visual Studio Code now auto-updates with TypeScript 2.4.1, so the warnings are hard to opt out of at this point.

**Comment by iRoachie** [comment]
@dannycochran This is already live

**Comment by dannycochran** [comment]
Ah I see -- I think the usage of "ViewStyleProp" is still not everywhere it could be.

Do you know if there's any open PR for giving this treatment to TextStyle, as well? Otherwise I can open one.

**Comment by iRoachie** [comment]
@dannycochran It's there for TextStyle as well.

**Comment by newyankeecodeshop** [comment]
@dannycochran Yes, TextProperties has been updated, but it's possible some custom components aren't extending it or re-defined `style`. What component is broken for you?

**Comment by dannycochran** [comment]
@iRoachie thanks, missed it as I was searching for "TextStyleProp" but I see it's just inlined into TextInput and Text. I created a PR that abstracts the typing and uses it in a few other places.

I was using three components where the ViewStyleProp wasn't in effect (ActivityIndicator, ScrollView, KeyboardAvoidingView), but there were several others.

Just submitted a PR:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17990

**Comment by hpmax00** [comment]
Is the PR merged?
I use 2.9.2, and have the save issue
image
and ScrollableTabView type of @types/react-native-scrollable-tab-view
image

**Pull request merged**